### PR TITLE
Some fixes to run VQE trainings + derivative wrt RBS parameter

### PR DESCRIPTION
**Pull request by MatteoRobbiati**
With this PR I am:

1. fixing a couple of things related to recent qibo releases (e.g. tensorflow backend's migration);
2. writing a script to run the VQE training;
3. computing the derivative of the loss function w.r.t. the parameter of RBS gates as well (there is a decomposition + a combination of a couple of original PSR to get the result).

The VQE training script can be run with a similar shellscript:
```
#!/bin/bash

# Define variables for the parameters
OUTPUT_FOLDER="test"
CIRCUIT_ANSAZ="hw_preserving"
NQUBITS=4
NLAYERS=2
NSHOTS=10000
BACKEND="qiboml"
PLATFORM="tensorflow"
OPTIMIZER="sgd"
OPTIMIZER_OPTIONS="{ \"optimizer\": \"Adagrad\", \"learning_rate\": 0.05, \"nmessage\": 1, \"nepochs\": 50 }"

# Run the Python script with the specified arguments
python train_vqes.py    --output_folder "$OUTPUT_FOLDER" --circuit_ansatz "$CIRCUIT_ANSAZ" \
                        --nqubits "$NQUBITS" --nlayers "$NLAYERS" --backend $BACKEND \
                        --platform $PLATFORM --optimizer $OPTIMIZER \
                        --optimizer_options "$OPTIMIZER_OPTIONS"  --nshots $NSHOTS
```

### CODE REVIEW

**Review by marekgluza — APPROVED** [review]
All my comments are optional, just some questions because I was curious

Probably just sttraightaway esolve most comments unless you will feel that it's worth some more time 

Thank you Matteo!

File: train_vqes.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+import json
+import logging
+
+import qibo
+
+from boostvqe import ansatze
+from boostvqe.boost import dbqa_vqe
+
+
+class SpecificWarningFilter(logging.Filter):
```
[the remark this answers is not in the document]
Comment (marekgluza) [reply train_vqes.py:11]: @Sam-XiaoyueLi

File: src/boostvqe/ansatze.py [hunk 2]
```diff
@@ -54,28 +54,30 @@ def hw_preserving(nqubits, nlayers=1):
...
+
 def compute_gradients(parameters, circuit, hamiltonian):
     """
     Compute gradients of circuit's parameters to check the problem trainability.
     The evaluated derivatives are the ones of the expectation of `hamiltonian`
     over the final state get running `circuit.execute` w.r.t. rotational angles.
 
     """
-    tf_backend = construct_backend("tensorflow")
+    tf_backend = construct_backend(backend="qiboml", platform="tensorflow")
```
[the remark this answers is not in the document]
Comment (MatteoRobbiati) [reply src/boostvqe/ansatze.py:80]: We are refactoring Qibo on backends and one of the modifications is that all the QML backends have been migrated to [Qiboml](https://github.com/qiboteam/qiboml).

File: src/boostvqe/boost.py [hunk 3]
```diff
@@ -128,7 +127,6 @@ def dbqa_vqe(
         set_backend(backend=backend, platform=platform)
     else:
         set_backend(backend=backend)
```
[the remark this answers is not in the document]
Comment (MatteoRobbiati) [reply src/boostvqe/boost.py:129]: Above (in your previous comment) it was qiboml because in that specific function we are calculating gradients. To do so, we are using the [Tensorflow GradientTape](https://www.tensorflow.org/guide/autodiff) mechanism and for doing that we need a tensorflow backend.

File: src/boostvqe/plotscripts.py [hunk 4]
```diff
@@ -65,7 +65,7 @@ def plot_loss(
     plt.figure(figsize=(10 * width, 10 * width * 6 / 8))
     plt.title(title)
 
-    for i in range(config["nboost"]):
+    for i in range(config["nboost"] + 1):
```
[the remark this answers is not in the document]
Comment (MatteoRobbiati) [reply src/boostvqe/plotscripts.py:68]: yes I think it is better to put `nboost = 0`!

File: src/boostvqe/training_utils.py [hunk 5]
```diff
@@ -319,3 +321,38 @@ def parameter_shift(
 
     circuit.set_parameters(original)
     return float(generator_eigenval * (forward - backward))
+
+
+def quantum_derivative_via_shifts(
```
[the remark this answers is not in the document]
Comment (MatteoRobbiati) [reply src/boostvqe/training_utils.py:326]: No, this is implemented here from scratch! It is just a function to say: if my gate is a rotation do the old psr, otherwise if it is RBS do this combination.

File: train_vqes.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (MatteoRobbiati) [reply train_vqes.py:11]: yeah, this is a possible way to shut down a specific warning message :)

File: train_vqes.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
+import argparse
```
[the remark this answers is not in the document]
Comment (MatteoRobbiati) [reply train_vqes.py:1]: It is the same function wrote by @Edoardo-Pedicillo. This main is just a wrapper done to run it using a bashscript. I think the docstrings should be written in the function directly.

File: train_vqes.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    parser.add_argument(
+        "--boost_frequency", type=int, default=10, help="Frequency of boosting calls."
+    )
+    parser.add_argument(
+        "--dbi_steps", type=int, default=1, help="Number of DBI iterations."
+    )
+    parser.add_argument(
+        "--store_h", action="store_true", help="Store Hamiltonian at each iteration."
+    )
+    parser.add_argument(
```
[the remark this answers is not in the document]
Comment (MatteoRobbiati) [reply train_vqes.py:84]: Fine for me!

File: src/boostvqe/training_utils.py [hunk 8]
```diff
@@ -319,3 +321,38 @@ def parameter_shift(
...
+        for g in circuit.queue:
+            if g != target_gate:
+                unrolled_rbs_circuit.add(g)
+            else:
+                unrolled_rbs_circuit.add(gates.H(target_qubits[0]))
+                unrolled_rbs_circuit.add(gates.CNOT(target_qubits[0], target_qubits[1]))
+                unrolled_rbs_circuit.add(gates.H(target_qubits[1]))
+                unrolled_rbs_circuit.add(gates.RY(target_qubits[0], parameter_value))
+                unrolled_rbs_circuit.add(gates.RY(target_qubits[1], -parameter_value))
+                unrolled_rbs_circuit.add(gates.H(target_qubits[1]))
```
Comment (renatomello) [inline src/boostvqe/training_utils.py:348]: I only saw this now, but you could remove these two :math:`H` gates and flip the sign of the second :math:`RY` from negative to positive and it is also a valid decomposition.

Comment (renatomello) [reply src/boostvqe/training_utils.py:348]: > I only saw this now, but you could remove these two H gates and flip the sign of the second R Y from negative to positive and it is also a valid decomposition.

I'll update the default decomposition in `qibo` to the one with less qubits.

Comment (MatteoRobbiati) [reply src/boostvqe/training_utils.py:348]: Yes, it's true! Don't think it is affecting that much the computation, though.
